File: Chagok/backend/app/api/party.py
"""
Party API Router - REST endpoints for party node management
007-lawyer-portal-v1: US1 (Party Relationship Graph)
"""

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session
from typing import Optional

from app.core.dependencies import (
    get_db,
    get_current_user_id,
    verify_case_read_access,
    verify_case_write_access,
    verify_internal_api_key,
    get_llm_service,
    get_metadata_store_service
)
from app.db.models import PartyType
from app.db.schemas import (
    PartyNodeCreate,
    PartyNodeUpdate,
    PartyNodeResponse,
    AutoExtractedPartyRequest,
    AutoExtractedPartyResponse
)
from app.services.party_service import PartyService
from app.domain.ports.llm_port import LLMPort
from app.domain.ports.metadata_store_port import MetadataStorePort

router = APIRouter(
    prefix="/cases/{case_id}/parties",
    tags=["parties"]
)


@router.post(
    "",
    response_model=PartyNodeResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new party node"
)
async def create_party(
    case_id: str,
    data: PartyNodeCreate,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id)
):
    """
    Create a new party node for a case.

    Requires write access to the case.
    """
    verify_case_write_access(case_id, db, user_id)

    service = PartyService(db)
    return service.create_party(case_id, data, user_id)


@router.get(
    "",
    response_model=dict,
    summary="Get all parties for a case"
)
async def get_parties(
    case_id: str,
    type: Optional[PartyType] = Query(None, description="Filter by party type"),
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id)
):
    """
    Get all party nodes for a case.

    Optionally filter by party type.
    Requires read access to the case.
    """
    verify_case_read_access(case_id, db, user_id)

    service = PartyService(db)
    parties = service.get_parties_for_case(case_id, type)

    return {
        "items": parties,
        "total": len(parties)
    }


@router.get(
    "/{party_id}",
    response_model=PartyNodeResponse,
    summary="Get a specific party node"
)
async def get_party(
    case_id: str,
    party_id: str,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id)
):
    """
    Get a specific party node by ID.

    Requires read access to the case.
    """
    verify_case_read_access(case_id, db, user_id)

    service = PartyService(db)
    return service.get_party(party_id)


@router.patch(
    "/{party_id}",
    response_model=PartyNodeResponse,
    summary="Update a party node"
)
async def update_party(
    case_id: str,
    party_id: str,
    data: PartyNodeUpdate,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id)
):
    """
    Update a party node.

    Requires write access to the case.
    """
    verify_case_write_access(case_id, db, user_id)

    service = PartyService(db)
    return service.update_party(party_id, data, user_id)


@router.delete(
    "/{party_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a party node"
)
async def delete_party(
    case_id: str,
    party_id: str,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id)
):
    """
    Delete a party node.

    Warning: This will also delete all relationships and evidence links
    associated with this party.

    Requires write access to the case.
    """
    verify_case_write_access(case_id, db, user_id)

    service = PartyService(db)
    service.delete_party(party_id, user_id)

    return None


# Graph endpoint (separate prefix from parties)
graph_router = APIRouter(
    prefix="/cases/{case_id}",
    tags=["parties"]
)


@graph_router.get(
    "/graph",
    response_model=dict,
    summary="Get the full party graph for a case"
)
async def get_graph(
    case_id: str,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id)
):
    """
    Get the complete party relationship graph for a case.

    Returns both party nodes and relationships in a single response
    for efficient React Flow rendering.

    Requires read access to the case.
    """
    verify_case_read_access(case_id, db, user_id)

    service = PartyService(db)
    return service.get_graph(case_id)


# ============================================
# 012-precedent-integration: T040 자동 추출 엔드포인트
# ============================================
@router.post(
    "/auto-extract",
    response_model=AutoExtractedPartyResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Save auto-extracted party from AI Worker"
)
async def auto_extract_party(
    case_id: str,
    data: AutoExtractedPartyRequest,
    db: Session = Depends(get_db),
    _: bool = Depends(verify_internal_api_key)
):
    """
    Save a party auto-extracted by AI Worker.

    - Checks for duplicate names (similarity threshold 0.8)
    - Minimum confidence threshold: 0.7
    - Requires internal API key (X-Internal-API-Key header)

    Returns:
        - id: New or existing party ID
        - is_duplicate: True if matched existing party
        - matched_party_id: Existing party ID if duplicate
    """
    # Confidence threshold check
    if data.extraction_confidence < 0.7:
        from fastapi import HTTPException
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Extraction confidence must be at least 0.7"
        )

    service = PartyService(db)
    # AI Worker에서 호출되므로 user_id 대신 "ai_worker" 사용
    return service.create_auto_extracted_party(case_id, data, "ai_worker")


# ============================================
# 019-party-extraction-prompt: 인물 관계도 재생성 엔드포인트
# ============================================
@router.post(
    "/regenerate",
    response_model=dict,
    summary="Regenerate party graph from fact summary"
)
async def regenerate_party_graph(
    case_id: str,
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
    llm_port: LLMPort = Depends(get_llm_service),
    metadata_store_port: MetadataStorePort = Depends(get_metadata_store_service)
):
    """
    사실관계 요약을 기반으로 인물 관계도를 재생성합니다.

    - 기존 자동 추출된 인물/관계는 유지됩니다 (이름 기반 병합)
    - 수동 추가된 인물은 보존됩니다
    - 사실관계 요약이 없으면 오류가 발생합니다

    Requires write access to the case.
    """
    verify_case_write_access(case_id, db, user_id)

    from app.services.party_extraction_service import PartyExtractionService

    extraction_service = PartyExtractionService(
        db,
        llm_port=llm_port,
        metadata_store_port=metadata_store_port
    )
    result = extraction_service.extract_from_fact_summary(
        case_id=case_id,
        user_id=user_id
    )

    return {
        "success": True,
        "message": "인물 관계도가 재생성되었습니다",
        "new_parties_count": result.new_parties_count,
        "merged_parties_count": result.merged_parties_count,
        "new_relationships_count": result.new_relationships_count,
        "total_persons": len(result.persons),
        "total_relationships": len(result.relationships)
    }

File: Chagok/frontend/src/lib/api/party.ts
/**
 * Party API client for LEH Lawyer Portal v1
 * User Story 1: Party Relationship Graph
 */

import { apiClient } from './client';
import type {
  PartyNode,
  PartyNodeCreate,
  PartyNodeUpdate,
  PartyRelationship,
  RelationshipCreate,
  RelationshipUpdate,
  PartyGraphData,
} from '@/types/party';

// ============================================================
// Party Node APIs
// ============================================================

interface PartyListResponse {
  items: PartyNode[];
  total: number;
}

/**
 * Get all parties for a case
 */
export async function getParties(caseId: string): Promise<PartyNode[]> {
  const response = await apiClient.get<PartyListResponse>(
    `/cases/${caseId}/parties`
  );
  if (!response.data) {
    throw new Error(response.error || '당사자 목록을 불러오는데 실패했습니다.');
  }
  return response.data.items;
}

/**
 * Get a single party by ID
 */
export async function getParty(caseId: string, partyId: string): Promise<PartyNode> {
  const response = await apiClient.get<PartyNode>(
    `/cases/${caseId}/parties/${partyId}`
  );
  if (!response.data) {
    throw new Error(response.error || '당사자 정보를 불러오는데 실패했습니다.');
  }
  return response.data;
}

/**
 * Create a new party
 */
export async function createParty(
  caseId: string,
  data: PartyNodeCreate
): Promise<PartyNode> {
  const response = await apiClient.post<PartyNode>(
    `/cases/${caseId}/parties`,
    data
  );
  if (!response.data) {
    throw new Error(response.error || '당사자를 추가하는데 실패했습니다.');
  }
  return response.data;
}

/**
 * Update a party
 */
export async function updateParty(
  caseId: string,
  partyId: string,
  data: PartyNodeUpdate
): Promise<PartyNode> {
  const response = await apiClient.patch<PartyNode>(
    `/cases/${caseId}/parties/${partyId}`,
    data
  );
  if (!response.data) {
    throw new Error(response.error || '당사자 정보를 수정하는데 실패했습니다.');
  }
  return response.data;
}

/**
 * Delete a party
 */
export async function deleteParty(caseId: string, partyId: string): Promise<void> {
  const response = await apiClient.delete(`/cases/${caseId}/parties/${partyId}`);
  if (response.error) {
    throw new Error(response.error || '당사자를 삭제하는데 실패했습니다.');
  }
}

// ============================================================
// Relationship APIs
// ============================================================

interface RelationshipListResponse {
  items: PartyRelationship[];
  total: number;
}

/**
 * Get all relationships for a case
 */
export async function getRelationships(caseId: string): Promise<PartyRelationship[]> {
  const response = await apiClient.get<RelationshipListResponse>(
    `/cases/${caseId}/relationships`
  );
  if (!response.data) {
    throw new Error(response.error || '관계 목록을 불러오는데 실패했습니다.');
  }
  return response.data.items;
}

/**
 * Create a new relationship
 */
export async function createRelationship(
  caseId: string,
  data: RelationshipCreate
): Promise<PartyRelationship> {
  const response = await apiClient.post<PartyRelationship>(
    `/cases/${caseId}/relationships`,
    data
  );
  if (!response.data) {
    throw new Error(response.error || '관계를 추가하는데 실패했습니다.');
  }
  return response.data;
}

/**
 * Update a relationship
 */
export async function updateRelationship(
  caseId: string,
  relationshipId: string,
  data: RelationshipUpdate
): Promise<PartyRelationship> {
  const response = await apiClient.patch<PartyRelationship>(
    `/cases/${caseId}/relationships/${relationshipId}`,
    data
  );
  if (!response.data) {
    throw new Error(response.error || '관계 정보를 수정하는데 실패했습니다.');
  }
  return response.data;
}

/**
 * Delete a relationship
 */
export async function deleteRelationship(
  caseId: string,
  relationshipId: string
): Promise<void> {
  const response = await apiClient.delete(
    `/cases/${caseId}/relationships/${relationshipId}`
  );
  if (response.error) {
    throw new Error(response.error || '관계를 삭제하는데 실패했습니다.');
  }
}

// ============================================================
// Graph API (combined data for React Flow)
// ============================================================

/**
 * Get the complete party graph for a case
 * Returns nodes and relationships in a single request
 */
export async function getPartyGraph(caseId: string): Promise<PartyGraphData> {
  const response = await apiClient.get<PartyGraphData>(
    `/cases/${caseId}/graph`
  );
  if (!response.data) {
    throw new Error(response.error || '관계도를 불러오는데 실패했습니다.');
  }
  return response.data;
}

// ============================================================
// Batch Position Update
// ============================================================

interface PositionUpdate {
  id: string;
  position: { x: number; y: number };
}

/**
 * Update multiple party positions in a single request
 * Used for persisting React Flow node positions
 */
export async function updatePartyPositions(
  caseId: string,
  positions: PositionUpdate[]
): Promise<void> {
  const response = await apiClient.put(
    `/cases/${caseId}/parties/positions`,
    { positions }
  );
  if (response.error) {
    throw new Error(response.error || '위치 저장에 실패했습니다.');
  }
}

// ============================================================
// 019-party-extraction-prompt: 인물 관계도 재생성 API
// ============================================================

export interface RegeneratePartyGraphResponse {
  success: boolean;
  message: string;
  new_parties_count: number;
  merged_parties_count: number;
  new_relationships_count: number;
  total_persons: number;
  total_relationships: number;
}

/**
 * Regenerate party graph from fact summary using AI
 * 사실관계 요약을 기반으로 인물 관계도를 재생성합니다.
 */
export async function regeneratePartyGraph(
  caseId: string
): Promise<RegeneratePartyGraphResponse> {
  const response = await apiClient.post<RegeneratePartyGraphResponse>(
    `/cases/${caseId}/parties/regenerate`
  );
  if (!response.data) {
    throw new Error(response.error || '인물 관계도 재생성에 실패했습니다.');
  }
  return response.data;
}
